cpp/src/arrow/filesystem/gcsfs.cc

```diff
@@ -33,13 +33,23 @@
 
 namespace arrow {
 namespace fs {
-struct GcsCredentials {
-  explicit GcsCredentials(std::shared_ptr<google::cloud::Credentials> c)
+struct GcsCredentialsHolder {
+  explicit GcsCredentialsHolder(std::shared_ptr<google::cloud::Credentials> c)
       : credentials(std::move(c)) {}
 
   std::shared_ptr<google::cloud::Credentials> credentials;
 };
 
+bool GcsCredentials::Equals(const GcsCredentials& other) const {
+  if (holder_->credentials == other.holder_->credentials) {
+    return true;
+  }
+  return anonymous_ == other.anonymous_ && access_token_ == other.access_token_ &&
```

Comment:
I am not sure I understand the motivation for these changes around credentials.  Is that to provide more "accurate" results from `operator==`?  If so, consider just adding a `debug_string_` or something similar, new credential types are possible in the future, and it seems sad to change a lot of code just to add a new credential type.

Ultimately this is your call, I do not understand the expectations in this project around values vs. opaque types.

Reply:
There seems to be an expectation of being able to serialize/deserialize the FS implementation via pickle.   I didn't see any method directly on credentials that would allow this, but happy to use both.  Since credentials didn't support this keeping all inputs seemed like the only viable way of doing it

Reply:
Ack. SGTM